### PR TITLE
SCP-2317 Convert cardano-api Tx types from/to our Tx types.

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Depends on #3335 being merged first.

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-contract/src/Plutus/Contract/CardanoAPI.hs [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+fromCardanoPlutusScript :: C.HasTypeProxy lang => C.PlutusScript lang -> P.Script
+fromCardanoPlutusScript = Codec.deserialise . BSL.fromStrict . C.serialiseToRawBytes
+
+toCardanoPlutusScript :: P.Script -> Either ToCardanoError (C.PlutusScript C.PlutusScriptV1)
+toCardanoPlutusScript = deserialiseFromRawBytes (C.AsPlutusScript C.AsPlutusScriptV1) . BSL.toStrict . Codec.serialise
+
+toCardanoExecutionUnits :: P.Script -> [Data.Data] -> Either ToCardanoError C.ExecutionUnits
+toCardanoExecutionUnits script datum = do
+    cmp <- maybe (Left NoDefaultCostModelParams) Right Api.defaultCostModelParams
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/CardanoAPI.hs:351]: Add a TODO to make the cost model params configurable

File: plutus-contract/src/Plutus/Contract/CardanoAPI.hs [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+fromCardanoTxInsCollateral (C.TxInsCollateral _ txIns) = Set.fromList $ fmap (P.pubKeyTxIn . fromCardanoTxIn) txIns
+
+toCardanoTxInsCollateral :: Set.Set P.TxIn -> Either ToCardanoError (C.TxInsCollateral C.AlonzoEra)
+toCardanoTxInsCollateral = fmap (C.TxInsCollateral C.CollateralInAlonzoEra) . traverse (toCardanoTxIn . P.txInRef) . Set.toList
+
+fromCardanoTxInWitness :: C.Witness C.WitCtxTxIn era -> Either FromCardanoError P.TxInType
+fromCardanoTxInWitness (C.KeyWitness C.KeyWitnessForSpending) = pure P.ConsumePublicKeyAddress
+fromCardanoTxInWitness
+    (C.ScriptWitness _
+        (C.PlutusScriptWitness C.PlutusScriptV1InAlonzo C.PlutusScriptV1
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/CardanoAPI.hs:123]: Are there no named fields for this?

**Review by j-mueller — APPROVED** [review]
Looks good. I suspect there may be a few errors that we only discover when we actually use the functions (for example in `toCardanoMintWitness` or computing hashes) but we can deal with them later.

**Comment by sjoerdvisscher** [comment]
🎉

**Pull request merged**